### PR TITLE
add Eureka TLS/mTLS support

### PR DESCRIPTION
**Pull request by julian-hj**
This pull request allows us to specify PEM encoded key and certificate values for communicating with a Eureka server that has mTLS enabled. We needed to convert those values into a truststore and keystore file in order to work with the spring-cloud-netflix service discovery client, so this PR also pulls over a copy of the [org.springframework.boot.ssl.pem](https://github.com/spring-projects/spring-boot/tree/main/spring-boot-project/spring-boot/src/main/java/org/springframework/boot/ssl/pem) package, dumbed down a bit and made Java 8 compatible.

- read `ca.crt`, `tls.crt` and `tls.key` optional properties in eureka bindings
- process them into keystore and truststore, and wire up eureka client/instance properties
- place the trust-store and keystore in the `$TMPDIR` and reference them by absolute path.
- add a `pem` package for parsing and converting PEM encoded RSA keys and certificates. (This package may eventually be replaced by the PEM package in spring-boot, but not until 3.1 is the minimum supported boot version. The package here is copied from spring-boot but simplified and made Java 8 compatible.)
- set `eureka.instance.preferIpAddress` to `true` when `ca.crt` is set, and the property doesn't already exist in the environment. This provides a working default when no property is set (since `$HOSTNAME` isn't a routable value in k8s) but allows applications to override the behavior if they want to set their own `eureka.instance.host` property.

### CODE REVIEW

File: src/main/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessor.java [hunk 1]
```diff
@@ -47,9 +55,82 @@ public void process(Environment environment, Bindings bindings, Map<String, Obje
...
+                    KeyStore ks = PemSslStoreHelper.createKeyStore("trust", "PKCS12", caCert, null, "rootca");
+                    FileOutputStream fos = new FileOutputStream(trustFilePath);
+                    ks.store(fos, generatedPassword.toCharArray());
+                    fos.close();
+                } catch (FileNotFoundException e) {
+                    throw new IllegalStateException("Unable to open truststore output file", e);
+                } catch (KeyStoreException e) {
+                    throw new IllegalStateException("Unable to write truststore", e);
+                } catch (NoSuchAlgorithmException e) {
+                    throw new IllegalStateException("Unable to write truststore", e);
```
Comment (anthonydahanne) [inline src/main/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessor.java:88]: well, no, it's rather that the cryptographic algorithm is not available

File: src/main/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessor.java [hunk 2]
```diff
@@ -47,9 +55,82 @@ public void process(Environment environment, Bindings bindings, Map<String, Obje
...
+                        .collect(StringBuilder::new, StringBuilder::appendCodePoint, StringBuilder::append)
+                        .toString();
+
+                // Create a trust store from the CA cert
+                String trustFilePath = Paths.get(System.getProperty("java.io.tmpdir"), "client-truststore.p12").toString();
+                try {
+                    KeyStore ks = PemSslStoreHelper.createKeyStore("trust", "PKCS12", caCert, null, "rootca");
+                    FileOutputStream fos = new FileOutputStream(trustFilePath);
+                    ks.store(fos, generatedPassword.toCharArray());
+                    fos.close();
```
Comment (anthonydahanne) [inline src/main/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessor.java:82]: hum, what if `ks.store` threw an exception? fos won't be closed ?

I suggest you reorganize try/catch blocks

Does that look suitable to you:

```java
                try {
                    FileOutputStream fos = new FileOutputStream(trustFilePath);
                    try {
                        ks.store(fos, generatedPassword.toCharArray());
                    } catch (KeyStoreException e) {
                        throw new IllegalStateException("Unable to write truststore", e);
                    } catch (NoSuchAlgorithmException e) {
                        throw new IllegalStateException("Cryptographic algorithm not available", e);
                    } catch (CertificateException e) {
                        throw new IllegalStateException("Unable to process certificate", e);
                    } catch (IOException e) {
                        throw new IllegalStateException("Unable to create truststore", e);
                    } finally {
                        try {
                            fos.close();
                        } catch (IOException e) {
                            throw new IllegalStateException("Unable to close truststore output file", e);
                        }
                    }
                } catch (FileNotFoundException e) {
                    throw new IllegalStateException("Unable to open truststore output file", e);
                }
```

File: src/main/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessor.java [hunk 3]
```diff
@@ -47,9 +55,82 @@ public void process(Environment environment, Bindings bindings, Map<String, Obje
...
+                } catch (IOException e) {
+                    throw new IllegalStateException("Unable to create truststore", e);
+                }
+                properties.put("eureka.client.tls.enabled", true);
+                properties.put("eureka.client.tls.trust-store", "file:"+trustFilePath);
+                properties.put("eureka.client.tls.trust-store-type", "PKCS12");
+                properties.put("eureka.client.tls.trust-store-password", generatedPassword);
+
+                // When tls.crt and tls.key are set, enable mTLS for Eureka
+                String clientKey = secret.get("tls.key");
```
Comment (anthonydahanne) [inline src/main/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessor.java:100]: this code is a duplicate of what you wrote above; please extract in a common method, like this:

```java
    private static void createStore(String generatedPassword, String filePath, KeyStore ks, String storeType) {
        try {
            FileOutputStream fos = new FileOutputStream(filePath);
            try {
                ks.store(fos, generatedPassword.toCharArray());
            } catch (KeyStoreException e) {
                throw new IllegalStateException("Unable to write " + storeType, e);
            } catch (NoSuchAlgorithmException e) {
                throw new IllegalStateException("Cryptographic algorithm not available", e);
            } catch (CertificateException e) {
                throw new IllegalStateException("Unable to process certificate", e);
            } catch (IOException e) {
                throw new IllegalStateException("Unable to create " + storeType, e);
            } finally {
                try {
                    fos.close();
                } catch (IOException e) {
                    throw new IllegalStateException("Unable to close " + storeType + " output file", e);
                }
            }
        } catch (FileNotFoundException e) {
            throw new IllegalStateException("Unable to open " + storeType + " output file", e);
        }
    }
```

File: src/main/java/org/springframework/cloud/bindings/boot/pem/package-info.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * SSL trust material provider for PEM-encoded certificates.
+ */
+package org.springframework.cloud.bindings.boot.pem;
```
Comment (anthonydahanne) [inline src/main/java/org/springframework/cloud/bindings/boot/pem/package-info.java:20]: please give references where this code comes from, and why we don't import the jars instead of copying / pasting

File: src/main/java/org/springframework/cloud/bindings/boot/pem/PemCertificateParser.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	 * @return the parsed certificates
+	 */
+	static X509Certificate[] parse(String certificates) {
+		if (certificates == null) {
+			return null;
+		}
+		CertificateFactory factory = getCertificateFactory();
+		List<X509Certificate> certs = new ArrayList<>();
+		readCertificates(certificates, factory, certs::add);
+		return (!certs.isEmpty()) ? certs.toArray(X509Certificate[]::new) : null;
```
Comment (anthonydahanne) [inline src/main/java/org/springframework/cloud/bindings/boot/pem/PemCertificateParser.java:61]: `certs.toArray` does not exist in Java 8.
and this module still supports Java 8.

You can rather use the `stream` api: `certs.stream().`

File: src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java [hunk 6]
```diff
@@ -16,48 +16,180 @@
...
 package org.springframework.cloud.bindings.boot;
 
-import org.junit.jupiter.api.DisplayName;
-import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.*;
 import org.springframework.cloud.bindings.Binding;
 import org.springframework.cloud.bindings.Bindings;
 import org.springframework.cloud.bindings.FluentMap;
+import org.springframework.core.io.ClassPathResource;
+import org.springframework.core.io.FileSystemResource;
```
Comment (anthonydahanne) [inline src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java:24]: unused import

File: src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java [hunk 7]
```diff
@@ -16,48 +16,180 @@
 
 package org.springframework.cloud.bindings.boot;
 
-import org.junit.jupiter.api.DisplayName;
-import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.*;
```
Comment (anthonydahanne) [inline src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java:19]: avoid wildcard

File: src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java [hunk 8]
```diff
@@ -16,48 +16,180 @@
...
 import org.springframework.mock.env.MockEnvironment;
 
+import java.io.File;
 import java.nio.file.Paths;
 import java.util.HashMap;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
 import static org.springframework.cloud.bindings.boot.EurekaBindingsPropertiesProcessor.TYPE;
+import static org.springframework.cloud.bindings.boot.TestHelper.resourceAsString;
```
Comment (anthonydahanne) [inline src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java:34]: unused import

File: src/test/java/org/springframework/cloud/bindings/boot/TestHelper.java [hunk 9]
```diff
@@ -0,0 +1,17 @@
+package org.springframework.cloud.bindings.boot;
```
Comment (anthonydahanne) [inline src/test/java/org/springframework/cloud/bindings/boot/TestHelper.java:1]: you can simply move this static method to your test class `EurekaBindingsPropertiesProcessorTest`

File: src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java [hunk 10]
```diff
@@ -16,48 +16,180 @@
...
 @DisplayName("Eureka BindingsPropertiesProcessor")
 final class EurekaBindingsPropertiesProcessorTest {
-
-    private final Bindings bindings = new Bindings(
+    private Bindings bindings = new Bindings(
             new Binding("test-name", Paths.get("test-path"),
                     new FluentMap()
                             .withEntry(Binding.TYPE, TYPE)
                             .withEntry("uri", "test-uri")
-                            .withEntry("client-id", "test-client-id")
```
Comment (anthonydahanne) [inline src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java:40]: why dont you keep them and test them in `    @DisplayName("contributes only base properties when no auth is set")` ?

File: src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java [hunk 11]
```diff
@@ -67,5 +199,4 @@ void disabled() {
 
         assertThat(properties).isEmpty();
     }
-
```
Comment (anthonydahanne) [inline src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java:70]: please add some tests that make the code throw exceptions.
What would happen? kill the whole app startup? or continue without TLS config?

**Review by anthonydahanne — CHANGES_REQUESTED** [review]
Hey there!
Let me know if you're ok with those suggestions!

File: src/test/java/org/springframework/cloud/bindings/boot/TestHelper.java (hunk 9 above)
Comment (julian-hj) [reply src/test/java/org/springframework/cloud/bindings/boot/TestHelper.java:1]: Hmm. This helper code is also used by a few different test classes in the pem package, so it seemed a bit messy to have a bunch of different test classes all calling a static method on one test class. We put it into a helper class to make it clearer that it is shared code. Is there a better pattern?

File: src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java (hunk 10 above)
Comment (julian-hj) [reply src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java:40]: There is a valid test case where the binding contains the URI, but no credentials. We wanted to test that case. The case where oauth2 credentials are set is in a separate, more explicit test now.

File: src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java (hunk 11 above)
Comment (julian-hj) [reply src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java:70]: Do you have suggestions on where that test code should live? To me, this unit test maybe isn't it. We've converted everything to a runtime exception, but maybe to your point, I am not sure what the calling code will do when it gets one of our IllegalStateExceptions.

My thought is that we probably *do* want the app to fail startup if we have a binding we cannot process, but I don't think we can test whether or not that will happen in this class.

File: src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java (hunk 7 above)
Comment (julian-hj) [reply src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java:19]: ok

File: src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java (hunk 8 above)
Comment (julian-hj) [reply src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java:34]: ok

File: src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java (hunk 6 above)
Comment (julian-hj) [reply src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java:24]: ok

File: src/main/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessor.java [hunk 12]
```diff
@@ -19,8 +19,16 @@
 import org.springframework.cloud.bindings.Binding;
 import org.springframework.cloud.bindings.Bindings;
 import org.springframework.core.env.Environment;
+import org.springframework.cloud.bindings.boot.pem.*;
```
[the remark this answers is not in the document]
Comment (julian-hj) [reply src/main/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessor.java:22]: OK. (we only actually used one class int this file so definitely didn't need the wildcard.

File: src/main/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessor.java [hunk 13]
```diff
@@ -47,9 +55,82 @@ public void process(Environment environment, Bindings bindings, Map<String, Obje
             map.from("uri").to("eureka.client.serviceUrl.defaultZone",
                     (uri) -> String.format("%s/eureka/", uri)
             );
-
             properties.put("eureka.client.region", "default");
-        });
 
+            Map<String, String> secret = binding.getSecret();
```
[the remark this answers is not in the document]
Comment (julian-hj) [reply src/main/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessor.java:60]: OK

File: src/main/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessor.java (hunk 1 above)
Comment (julian-hj) [reply src/main/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessor.java:88]: ok

File: src/main/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessor.java (hunk 2 above)
Comment (julian-hj) [reply src/main/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessor.java:82]: This makes sense.

File: src/main/java/org/springframework/cloud/bindings/boot/pem/package-info.java (hunk 4 above)
Comment (julian-hj) [reply src/main/java/org/springframework/cloud/bindings/boot/pem/package-info.java:20]: I took a crack at this. Will re-push shortly with that change.

File: src/main/java/org/springframework/cloud/bindings/boot/pem/PemCertificateParser.java (hunk 5 above)
Comment (julian-hj) [reply src/main/java/org/springframework/cloud/bindings/boot/pem/PemCertificateParser.java:61]: I'm a bit puzzled here, but if you couldn't tell, I am not a Java person, so I am sure I'm missing something. From what I can tell, ArrayList.toArray() is supported in Java 8: https://docs.oracle.com/javase/8/docs/api/java/util/ArrayList.html#toArray--

I also don't see a `stream` method on ArrayList.

Comment (anthonydahanne) [reply src/main/java/org/springframework/cloud/bindings/boot/pem/PemCertificateParser.java:61]: > ArrayList.toArray() is supported in Java 8

yes it is; but you declared `certs` as a `List` (not `ArrayList` which is concrete type), which inherits from `Collection` which only recently (well, Java `11`) added `toArray(IntFunction<T[]> generator)(IntFunction<T[]> generator` (https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Collection.html#toArray(java.util.function.IntFunction))

> I also don't see a stream method on ArrayList.
yes, it's on `Collection` as well (https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Collection.html#stream())

this compiles fine on Java 8:

```java
return (!certs.isEmpty()) ? certs.stream().toArray(X509Certificate[]::new) : null;
```

File: src/test/java/org/springframework/cloud/bindings/boot/TestHelper.java (hunk 9 above)
Comment (anthonydahanne) [reply src/test/java/org/springframework/cloud/bindings/boot/TestHelper.java:1]: well, that's fine, it's just for a test. When I wrote this I probably did not pay attention to the `pem` package usage.

You can leave as is.

File: src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java (hunk 10 above)
Comment (anthonydahanne) [reply src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java:40]: > The case where oauth2 credentials are set is in a separate, more explicit test now.

true, I did not notice; in `@DisplayName("contributes oauth properties when set")`

All good then!

File: src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java (hunk 11 above)
Comment (anthonydahanne) [reply src/test/java/org/springframework/cloud/bindings/boot/EurekaBindingsPropertiesProcessorTest.java:70]: ok, let me rephrase my suggestion.

I would like at least 1 test that fails to create a store and throws one of those `IllegalStateException`.

This is part of your changes, so it definitely makes sense to add it as part of your tests - for example, re do the test  named `@DisplayName("contributes mTLS properties when set")`  but this time with a `cert` and /or `key` like:

```java
            this.cert = "jklkl";
            this.key = "lkjlkjkl";
```

You'll be able to assert that `IllegalException` are supposed to be thrown in those cases.

---
> My thought is that we probably do want the app to fail startup if we have a binding we cannot process, but I don't think we can test whether or not that will happen in this class.

About integrations tests, we have [some](https://github.com/spring-cloud/spring-cloud-bindings/pull/94/files#diff-5fdfc03b2902a383ceb5db7636ec92877c04b81f29539c553ef92c15f4716003R10) that exist in a PR that can't be merged yet; it could be interesting to add an `Eureka` integration test that would break the startup of the app.

**Review by anthonydahanne — APPROVED** [review]
all tests are reviewed and passing.
README updated.
